File: rap-extensions/bundles/org.eclipse.rap.rwt/js/rwt/remote/RemoteObject.js
/*******************************************************************************
 * Copyright (c) 2012, 2014 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    EclipseSource - initial API and implementation
 ******************************************************************************/

(function(){

var Connection = rwt.remote.Connection;

/**
 * @private
 * @class Instances of RemoteObject represent the server-side counterpart of a client object
 * and are used to write operations into the next protocol message.
 * @description The constructor is not public. Instances can
 * be obtained from {@link rap.getRemoteObject}.
 * @exports rwt.remote.RemoteObject as RemoteObject
 * @since 2.0
 * @param {}
 *
 */
rwt.remote.RemoteObject = function( id ) {
  this._ = {
    "id" : id,
    "listen" : {}
  };
};

rwt.remote.RemoteObject.prototype = {

  /**
   * @description Sets the specified property of the remote object to the given value.
   * Calling this method multiple times for the same property will overwrite the previous value,
   * the message will not become longer. This method does not cause the message to be sent
   * immediately. Instead it will be sent the next time a "notify" or "call" operation is
   * written to the message.
   * @param {string} property The name of the property.
   * @param {var} value The value of the property.
   */
  set : function( key, value ) {
    Connection.getInstance().getMessageWriter().appendSet( this._.id, key, value );
  },

  /**
   * @description Notifies the remote object that an event of the given type occurred.
   * Notifications can only be sent for types that the server is currently listening for
   * (see {@link rap.registerTypeHandler}, <b>handler.events</b>). If this is not the
   * case, no "notify" operation is written into the message and no request will be sent.
   * Otherwise the message will be sent to the server within a few milliseconds. One message
   * may contain several "notify" operations, if they are added consecutively.
   * @param {string} event The type of the event that occured.
   * @param {Object|null} [properties] This object may contain any number of additional
   * properties/fields associated with the event. It may also be null or omitted.
   */
  notify : function( event, properties ) {
    var suppressSend = arguments[ 2 ];
   // TODO [tb]: suppressSend (or something similar) should be public API
    var actualProps = properties ? properties : {};
    if( this.isListening( event ) ) {
      var connection = Connection.getInstance();
      connection.getMessageWriter().appendNotify( this._.id, event, actualProps );
      if( suppressSend !== true ) {
        if( typeof suppressSend === "number" ) {
          connection.sendDelayed( suppressSend );
        } else {
          connection.send();
        }
      }
    }
  },

  /**
   * @description Instructs the remote object to call the given method.
   * Calling this method will write a "call" operation into the message, which will to be sent to
   * the server within a few milliseconds. One message
   * may contain several "call" operations, if they are added consecutively.
   * @param {string} method The name of the method.
   * @param {Object|null} [properties] This object may contain any number of additional
   * properties/fields associated with the call. It may also be null or omitted.
   */
  call : function( method, properties ) {
    var actualProps = properties ? properties : {};
    Connection.getInstance().getMessageWriter().appendCall( this._.id, method, actualProps );
    Connection.getInstance().send();
  },

  isListening : function( type ) {
    return this._.listen[ type ] === true;
  }

};

}());

File: rap-extensions/bundles/org.eclipse.rap.rwt/src/org/eclipse/rap/rwt/remote/RemoteObject.java
/*******************************************************************************
 * Copyright (c) 2011, 2013 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    EclipseSource - initial API and implementation
 ******************************************************************************/
package org.eclipse.rap.rwt.remote;

import org.eclipse.rap.json.JsonObject;
import org.eclipse.rap.json.JsonValue;


/**
 * Instances of this interface represent an object on the client that is synchronized with an object
 * on the server. A RemoteObject is used to transfer changes to the remote object. To create a
 * remote object, use {@link Connection#createRemoteObject(String)}.
 *
 * @noimplement This interface is not intended to be implemented by clients.
 * @see Connection
 * @since 2.0
 */
public interface RemoteObject {

  /**
   * Returns the id that is used to synchronize this object with the client. This id is created by
   * the framework and does not change over the lifetime of the object.
   *
   * @since 2.1
   */
  String getId();

  /**
   * Sets the specified property of the remote object to the given value.
   *
   * @param name the name of the property to set
   * @param value the value to set
   */
  void set( String name, int value );

  /**
   * Sets the specified property of the remote object to the given value.
   *
   * @param name the name of the property to set
   * @param value the value to set
   */
  void set( String name, double value );

  /**
   * Sets the specified property of the remote object to the given value.
   *
   * @param name the name of the property to set
   * @param value the value to set
   */
  void set( String name, boolean value );

  /**
   * Sets the specified property of the remote object to the given value.
   *
   * @param name the name of the property to set
   * @param value the value to set
   */
  void set( String name, String value );

  /**
   * Sets the specified property of the remote object to the given value.
   *
   * @param name the name of the property to set
   * @param value the value to set, must not be <code>null</code>
   * @since 2.1
   */
  void set( String name, JsonValue value );

  /**
   * Instructs the remote object to listen or to stop listening on the given type
   * of events. When the client is listening, it must notify the server when an
   * event of the given type occurs.
   *
   * @param eventType the name of event type to listen
   * @param listen true to listen to this type of events, false otherwise
   */
  void listen( String eventType, boolean listen );

  /**
   * Calls the method with the given name on the remote object.
   *
   * @param method the name of the method to call
   * @param parameters the named parameters to pass to the method call
   * @since 2.1
   */
  void call( String method, JsonObject parameters );

  /**
   * Instructs the client to destroy the remote object.
   */
  void destroy();

  /**
   * Set a handler to process incoming operations from the remote object.
   *
   * @param handler the handler that processes incoming operation
   */
  void setHandler( OperationHandler handler );

}
